### ISSUE TITLE
dataset attach memory issues

### ISSUE THREAD

**Issue by tmorrell** [caltechlibrary/dataset#104]
The dataset attach command for a 120GB file causes very high memory use (at least 29 GB) that doesn't seem to decrease as the files are transferred. The command is slow and can result in crashes. It impacts both the command line tool and shared library. As this is just moving files on the file system, there must be a more memory efficient approach.

**Comment by rsdoiel** [comment]
If I remember correctly I am reading the file into a dynamic memory buffer
then writing it out. This is why the memory usage initially blooms is
caused by this approach (which I can change). The Garbage collector is
suposed to release the memory afterwards (sound like it is not).

I can take a different approach in copying files using a fixed sized buffer
and many writes. That should solve the issue of ever expanding memory
consumption.
I'll try to get on this later today, it'll take a while to complete as I
need to develop a number of tests to confirm correctness in
my implementation.

Thanks,

Robert

**Comment by rsdoiel** [comment]
After reviewing the code I currently read in the content in order to calculate a checksum (a md5 hash of the content).
I need to figure out how to do that same thing without reading the whole file.  I have added an explicit `content = nil` to get the GC to release the memory once I use it.  Still working on alternative approach.

**Comment by tmorrell** [comment]
The fix definitely solved the memory issue for the dataset application, and I think (but am not 100% sure) it also solved the issue for the shared library.

However, I am seeing a bug where I get a 'key not found' error when I try to attach files to a newly created record. Dataset read can return back the json for the new key, but attach afterwards with the same key results in an error. Attach works for the same key once I start a new python session.

**Comment by rsdoiel** [comment]
OK, this sorta makes sense.  Right now I think I'm maintaining a list of
keys for a collection in memory to save a disc reads (e..g has_keys(),
keys()). I think I am doing that at the libdataset.go level. If so either I
can probably drop that "efficiency" or make it attachment aware.  I'll fix
that and then do another pre-release.

Thanks,

Robert

**Comment by rsdoiel** [comment]
Hello Tom,  finally got back to look at this. Maybe I am not understanding what remains broken.
Does the following work for you? (I'm assuming you are in a directory with "README.md").

```
from py_dataset import dataset
c_name = 't3.ds'
key = 'a1'
obj = {'one':key}
f_name = 'README.md'
dataset.init(c_name)
dataset.create(c_name, key, obj)
keys = dataset.keys(c_name)
print(keys)
o, err = dataset.read(c_name, key)
print(o)
dataset.attach(c_name, key, [ f_name ])
dataset.attachments(c_name, key)
o , err = dataset.read(c_name, key)
print(o)
```

**Comment by rsdoiel** [comment]
Wait, just spotted some documentation issues.  I think the trouble is the mapping of the args. It is wrong in the docs.

```
v2 = 'v2.0.0'
dataset.attach(c_name, key, [ f_name ], v2)
```

Is documented as 

```
dataset.attach(c_name, key, v2, [ f_name ])
```

**Comment by tmorrell** [comment]
So I've looked at this, and it's even weirder. The problem is because I'm changing directories during the operation. Here is some example code. The first attach works, the second doesn't. If you run the exact same script again (on a collection where the keys are already there), both attachments work. If you delete os.chdir('../') and os.chdir(temp) (and get the paths right), it will work.

This is probably some obscure python behavior. I'll probably just change my code so it doesn't change directories so much. But it's just so weird...

```import os
from py_dataset import dataset
c_name = 't3.ds'
key = 'a1'
obj = {'one':key}
f_name = 'README.md'
dataset.init(c_name)
dataset.create(c_name, key, obj)
keys = dataset.keys(c_name)
print(keys)
o, err = dataset.read(c_name, key)
print(o)
temp = 'temp'
os.chdir(temp)
dataset.attach('../'+c_name, key, [ f_name ])
dataset.attachments('../'+c_name, key)
o , err = dataset.read('../'+c_name, key)
print(o)
key = 'a2'
obj = {'two':key}
os.chdir('../')
create = dataset.create(c_name, key, obj)
print('CREATE',create)
keys = dataset.keys(c_name)
os.chdir(temp)
print(keys)
print(os.getcwd())
result = dataset.attach('../'+c_name, key, [ f_name ])
print(result, dataset.error_message())
o , err = dataset.read(c_name, key)
print(o)
```

**Comment by rsdoiel** [comment]
This is starting to make sense. Here's what I think may be happening.  The
libdataset shared library instance is running in its own thread/process. At
the point it is loaded (probably on the first open of the dataset
collection) it would inherit the "working directory" from the Python3
environment.  When you change the directory via os.chdir() the libdataset
thread is completely unaware of the change so all the paths you're passing
to it don't make sense any more.  Assuming I am correct we could add two
methods to libdataset to get/set the current working directory then in the
Python wrapper make calls to those before making the C API call the
libdataset library. We're making things pretty complicated at if we go this
route.

I am wondering if writing a pure python version of dataset make sense at
this point in time.  It might be slightly slower but it certainly would
make releases and updates easier.
We could use Cython to create a common line version. I think our
COLLECTION.ds/collection.json file structure is very stable now, I would
skip the Namaste stuff as that hasn't proven useful in my opinion. I would
definitely continue to include codameta.json along side the collection.json
as that is much more useful then the namaste directories.

I'm not sure the Go speed up is buying us much today. I don't build Go web
services for library projects anymore.  We're not using Go for a common
abstraction for storage (e.g. S3 support). As long as we use a pairtree,
handle the semver directories for attachments, I think we can nearly the
same performance using pure python and we could just let the Go version go
(pun intended).

What do you think?

Thanks,

Robert

**Comment by tmorrell** [comment]
I agree that getting and setting the working directory is too complicated.

A pure python version of dataset could make sense. The only performance question for me is the frames generation. If that's as efficient in Python, we should be good. 

I'm just going to change my code to avoid doing directory changes for the moment though :)

**Comment by rsdoiel** [comment]
I think frames will likely have a similar performance profile. I use a
pretty simple set of structure and we would be skipping the to/from JSON
encoding between Python space and C API space. Might be faster as pure
python.  We should get Mike's input on this.  Having everyone in our group
able to improve dataset would be cool.

- Robert

**Comment by mhucka** [comment]
Just chiming in – the analysis about the different address spaces being behind Tom's problem seems right. I know interfacing to native code from Python can lead to weird interactions, and especially if the different parts (Python interpreter, native code) keep state separately.

I don't know what the guts of dataset do, so can't say how much writing in pure Python would impact performance. We could be careful about the Python implementation to get as much performance as possible (doing things like profiling the code, which is pretty easy to do in Python). But will it ever be as fast as the Go implementation? Hard to say without trying it.

**Comment by rsdoiel** [comment]
I think doing a pure Python 3 implementation may have significant windfalls
aside from performance. I suspect IO will be the bounding factor in speed
so I am not sure at this stage Go is bringing much to the table, there are
also libev like libraries (twisted framework comes to mind from back in the
old days) we could leverage that sorta of things if IO is an issue. I'm not
really sure memory wise whether Python 3 or Go would be better (I would
assume Go given where it grew up, but that's not necessarily where Go is
today, Go seems to be accruing features at a very rapid rate).

The three pluses I see with a pure Python version:

+  hackability, we can all hack on it so it improves faster and becomes
more robust
+ easier adoption at Caltech (not sure this is really a goal but there
seems to be a lot of Python at Caltech, not so much for Go)
+ The Library/Archival communities seem to be open to Python, not sure
about Go
+ given how py_dataset is structured we can gradually move to a pure Python
solution, e.g. replacement of libdataset function by function

I'm not banging on Go just to bang of Go, I like coding in Go and for me
I'll probably still write Go for sometime even after Go gets bloated with
Go 2.x.  But I really think we need ensemble programming options and for
DLD Python 3 seems the strongest candidate.

Anyway, just my random thinking, there is no rush,

Robert

**Comment by mhucka** [comment]
Taking a quick look around, it looks like there are a number of Python i/o options (and an implementation of memory mapping too, though I think it's likely to work only on *nix and Windows) that are said to offer good performance, so we may be good on this front after all.

I'm certainly in favor of Python, but I'm biased :)

**Comment by rsdoiel** [comment]
I'm going to close this. The web service implementation, datasetd, will have features (e.g. frames) grow over time to replicate the command line version. This avoids the whole parallel issue of memory management using the Go generated C-Shared Library approach in Python 3.  

Additionally the structure of colllection.json is reasonably stable and the pairtree approach to handline JSON documents and attachments has been very stable for a while. Implementing a pure Python version makes the most sense from a python integration point of view. The only additional feature would be to respect the "lock.pid" file in the collection root that was added in 1.1.0 of dataset and datasetd.
